File: src/deadlock_build_sync/purchase_guidance_types.py
"""Typed purchase guidance shared by evidence, generation, and review."""

from __future__ import annotations

from dataclasses import asdict, dataclass
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .match_choices import AutomaticBranch


@dataclass(frozen=True)
class PurchaseTiming:
    item_id: int
    buyers: int
    counts_by_checkpoint: tuple[int, ...]

    @property
    def observed_position(self) -> int:
        return max(
            range(len(self.counts_by_checkpoint)),
            key=self.counts_by_checkpoint.__getitem__,
        )

    @property
    def support(self) -> int:
        return self.counts_by_checkpoint[self.observed_position]

    @property
    def position(self) -> int | None:
        if self.support >= 20 and self.support >= self.buyers * 0.1:
            return self.observed_position
        return None


@dataclass(frozen=True)
class ItemPurpose:
    label: str
    trigger: str
    evidence: str
    basis: str


@dataclass(frozen=True)
class PurchaseStep:
    item_id: int
    name: str
    incremental_cost: int
    cumulative_cost: int
    consumed_items: tuple[int, ...]
    owned_after: tuple[int, ...]


@dataclass(frozen=True)
class PurchasePlan:
    actions: tuple[PurchaseStep, ...]
    final_inventory: tuple[int, ...]
    remaining_cost: int
    decision: str
    save_souls: int | None


@dataclass(frozen=True)
class PurchaseState:
    owned: tuple[int, ...] = ()
    liquid_souls: int | None = None
    flex: int = 0


@dataclass(frozen=True)
class PurchaseChoice:
    item_id: int
    name: str
    tier: int
    catalog_cost: int
    purpose: ItemPurpose
    after_step: int | None
    timing: PurchaseTiming | None
    timing_basis: str
    route: tuple[int, ...]
    upgrades_core: tuple[int, ...]
    plan: PurchasePlan | None
    blocked_reason: str | None
    extra_path_cost: int | None
    rebought_components: tuple[int, ...]


@dataclass(frozen=True)
class PurchaseDecision:
    after_step: int
    kind: str
    purpose: str
    options: tuple[int, ...]
    upgrade_fork: bool = False


@dataclass(frozen=True)
class PurchaseGuidance:
    core_ids: tuple[int, ...]
    default_path: PurchasePlan
    choices: tuple[PurchaseChoice, ...]
    decisions: tuple[PurchaseDecision, ...]
    names: dict[int, str]
    evidence_basis: str = "Admitted production core; optional effects and timing do not prove an outcome benefit"
    schema_version: int = 2
    automatic_branches: tuple[AutomaticBranch, ...] = ()

    def as_dict(self) -> dict[str, object]:
        result = asdict(self)
        result["names"] = {str(item): name for item, name in self.names.items()}
        return result

File: crates/deadlock-guides/src/purchase_guidance_types.rs
use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::automatic_branch::AutomaticBranch;
use crate::purchase_plan_types::PurchasePlan;
use crate::purchase_timing::PurchaseTiming;

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ItemPurpose {
    pub label: String,
    pub trigger: String,
    pub evidence: String,
    pub basis: String,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct PurchaseChoice {
    pub item_id: u64,
    pub name: String,
    pub tier: u8,
    pub catalog_cost: u64,
    pub purpose: ItemPurpose,
    pub after_step: Option<usize>,
    pub timing: Option<PurchaseTiming>,
    pub timing_basis: String,
    pub route: Vec<u64>,
    pub upgrades_core: Vec<u64>,
    pub plan: Option<PurchasePlan>,
    pub blocked_reason: Option<String>,
    pub extra_path_cost: Option<i128>,
    pub rebought_components: Vec<u64>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct PurchaseDecision {
    pub after_step: usize,
    pub kind: String,
    pub purpose: String,
    pub options: Vec<u64>,
    pub upgrade_fork: bool,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct PurchaseGuidance {
    pub core_ids: Vec<u64>,
    pub default_path: PurchasePlan,
    pub choices: Vec<PurchaseChoice>,
    pub decisions: Vec<PurchaseDecision>,
    pub names: BTreeMap<u64, String>,
    pub evidence_basis: String,
    pub schema_version: u8,
    pub cohort: Value,
    pub evidence: Value,
    pub automatic_branches: Vec<AutomaticBranch>,
}
